File: CourseWork/LexicalAnalysis/Token.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;

namespace CourseWork.LexicalAnalysis
{
    public class Token
    {
        public Lexeme ParrentLexeme { get; }

        public TokenType Type { get; internal set; }
        public string File { get; }
        public int Line { get; }
        public int CharIndex { get; }

        public string StringValue { get; }
    
        private string _shortName;


        private static readonly  Dictionary<string, TokenType> _dictionary = new Dictionary<string, TokenType>()
        {
            { "[",       TokenType.Symbol },
            { "]",       TokenType.Symbol },
            { ".",       TokenType.Symbol },
            { ",",       TokenType.Symbol },
            { "+",       TokenType.Symbol },
            { "-",       TokenType.Symbol },
            { ":",       TokenType.Symbol },

            { "eax",     TokenType.Register32 },
            { "ebx",     TokenType.Register32 },
            { "ecx",     TokenType.Register32 },
            { "edx",     TokenType.Register32 },

            { "ax",      TokenType.Register16 },
            { "bx",      TokenType.Register16 },
            { "cx",      TokenType.Register16 },
            { "dx",      TokenType.Register16 },
            { "sp",      TokenType.Register16 },
            { "bp",      TokenType.Register16 },
            { "si",      TokenType.Register16 },
            { "di",      TokenType.Register16 },

            { "al",      TokenType.Register8 },
            { "cl",      TokenType.Register8 },
            { "dl",      TokenType.Register8 },
            { "bl",      TokenType.Register8 },
            { "ah",      TokenType.Register8 },
            { "ch",      TokenType.Register8 },
            { "dh",      TokenType.Register8 },
            { "bh",      TokenType.Register8 },

            { "es",      TokenType.SegmentRegister },
            { "cs",      TokenType.SegmentRegister },
            { "ss",      TokenType.SegmentRegister },
            { "ds",      TokenType.SegmentRegister },
            { "fs",      TokenType.SegmentRegister },
            { "gs",      TokenType.SegmentRegister },

            { "db",      TokenType.DbDirective },
            { "dw",      TokenType.DwDirective },
            { "dd",      TokenType.DdDirective },
            { "equ",     TokenType.EquDirective },
            { "if",      TokenType.IfDirective },
            { "else",    TokenType.ElseDirective },
            { "endif",   TokenType.EndifDirective },

            { "movsb",   TokenType.Instruction },
            { "pop",     TokenType.Instruction },
            { "dec",     TokenType.Instruction },
            { "add",     TokenType.Instruction },
            { "cmp",     TokenType.Instruction },
            { "bt",      TokenType.Instruction },
            { "jbe",     TokenType.Instruction },
            { "mov",     TokenType.Instruction },
            { "or",      TokenType.Instruction },

            { "segment", TokenType.SegmentKeyword},
            { "end",     TokenType.EndKeyword},
            { "ends",    TokenType.EndsKeyword},
        };

        private static readonly Regex _numberHexRegex = new Regex(@"^[0-9a-f]+h$");
        private static readonly Regex _numberDecRegex = new Regex("^[0-9]+$");
        private static readonly Regex _numberBinRegex = new Regex("^[01]+b$");
        private static readonly Regex _identifierRegex = new Regex(@"^[a-z]\w*$");

        public Token(string stringValue, string file, int line, int charIndex, Lexeme lexeme, out Error error)
        {
            error = null;
            ParrentLexeme = lexeme;
            StringValue = stringValue;

            Line = line;
            CharIndex = charIndex;
            if (file != null && System.IO.File.Exists(file))
            {
                File = file;
                _shortName = new FileInfo(File).Name;
            }

            if (!_dictionary.TryGetValue(stringValue, out var type))
            {
                if (_numberHexRegex.IsMatch(stringValue))
                {
                    Type = TokenType.HexNumber;
                }
                else if (_numberDecRegex.IsMatch(stringValue))
                {
                    Type = TokenType.DecNumber;
                }
                else if (_numberBinRegex.IsMatch(stringValue))
                {
                    Type = TokenType.BinNumber;
                }
                else if (stringValue.Trim().StartsWith("\"") && stringValue.Trim().EndsWith("\""))
                {
                    Type = TokenType.Text;
                }
                else if(_identifierRegex.IsMatch(stringValue))
                {
                    Type = TokenType.Identifier;
                }
                else
                {
                    Type = TokenType.Unknown;
                    error = new Error(ErrorType.UnknownToken, this);
                    return;
                }
            }
            else
            {
                Type = type;
            }

            if(Type == TokenType.Text)
                StringValue = stringValue.Trim(' ', '\t', '\"');
        }

        public bool HasValue()
        {
            return
                Type == TokenType.Instruction ||
                Type == TokenType.Register32 ||
                Type == TokenType.Register16 ||
                Type == TokenType.Register8 ||
                Type == TokenType.Symbol ||
                Type == TokenType.Identifier ||
                Type == TokenType.HexNumber ||
                Type == TokenType.DecNumber ||
                Type == TokenType.BinNumber ||
                Type == TokenType.Text ||
                Type == TokenType.UserSegment ||
                Type == TokenType.SegmentRegister ||
                Type == TokenType.Label;
        }

        public bool IsNumber()
        {
            return
                Type == TokenType.HexNumber ||
                Type == TokenType.DecNumber ||
                Type == TokenType.BinNumber;
        }

        public int ToValue()
        {
            switch (Type)
            {
                case TokenType.HexNumber:
                    return Convert.ToInt32(StringValue.Substring(0, StringValue.Length - 1), 16);
                case TokenType.DecNumber:
                    return Convert.ToInt32(StringValue, 10);
                case TokenType.BinNumber:
                    return Convert.ToInt32(StringValue.Substring(0, StringValue.Length - 1), 2);
                default:
                    throw new NotSupportedException();
            }
        }

        public override string ToString()
        {
            var sb = new StringBuilder();
            sb.AppendFormat("{0} ", Type);
            if (HasValue())
            {
                sb.AppendFormat("({0}) ", StringValue);
                if (IsNumber())
                    sb.AppendFormat("({0}) ", ToValue());
            }

            if (File != null)
                sb.AppendFormat("in {0} at {1}:{2}", _shortName, Line, CharIndex);
            else
                sb.AppendFormat("at {0}:{1}", Line, CharIndex);

            return sb.ToString();
        }

        public string ToSourceValue(bool dedStyle)
        {
            if (dedStyle)
            {
                return $"({"\"" + StringValue + "\"", -5} : {Type} : {StringValue.Length})";
            }
            else
            {
                if (HasValue())
                {
                    return IsNumber() ? ToValue().ToString() : StringValue;
                }

                return Type.ToString();
            }
        }
    }
}

File: java/src/main/java/course/work/sp/Token.java
package course.work.sp;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;


public class Token {

    public static final Map<String, TokenType> NAME_TO_TOKEN_TYPE_MAP;
    static {
        NAME_TO_TOKEN_TYPE_MAP = new HashMap<String, TokenType>();
        NAME_TO_TOKEN_TYPE_MAP.put("SEGMENT", TokenType.SegmentWord);
        NAME_TO_TOKEN_TYPE_MAP.put("DB", TokenType.DbDir);
        NAME_TO_TOKEN_TYPE_MAP.put("DW", TokenType.DwDir);
        NAME_TO_TOKEN_TYPE_MAP.put("DD", TokenType.DdDir);
        NAME_TO_TOKEN_TYPE_MAP.put("ASSUME", TokenType.KeyWord);
        NAME_TO_TOKEN_TYPE_MAP.put("STC", TokenType.Instruction);
        NAME_TO_TOKEN_TYPE_MAP.put("PUSH", TokenType.Instruction);
        NAME_TO_TOKEN_TYPE_MAP.put("JMP", TokenType.Instruction);
        NAME_TO_TOKEN_TYPE_MAP.put("JNC", TokenType.Instruction);
        NAME_TO_TOKEN_TYPE_MAP.put("MOV", TokenType.Instruction);
        NAME_TO_TOKEN_TYPE_MAP.put("MUL", TokenType.Instruction);
        NAME_TO_TOKEN_TYPE_MAP.put("XOR", TokenType.Instruction);
        NAME_TO_TOKEN_TYPE_MAP.put("BTR", TokenType.Instruction);
        NAME_TO_TOKEN_TYPE_MAP.put("SUB", TokenType.Instruction);
        NAME_TO_TOKEN_TYPE_MAP.put("ADC", TokenType.Instruction);
        NAME_TO_TOKEN_TYPE_MAP.put("END", TokenType.EndWord);
        NAME_TO_TOKEN_TYPE_MAP.put("ENDS", TokenType.EndsWord);
        NAME_TO_TOKEN_TYPE_MAP.put("[", TokenType.Symbol);
        NAME_TO_TOKEN_TYPE_MAP.put("]", TokenType.Symbol);
        NAME_TO_TOKEN_TYPE_MAP.put("+", TokenType.Symbol);
        NAME_TO_TOKEN_TYPE_MAP.put(":", TokenType.Symbol);
        NAME_TO_TOKEN_TYPE_MAP.put(",", TokenType.Koma);
        NAME_TO_TOKEN_TYPE_MAP.put("DWORD", TokenType.DwordWord);
        NAME_TO_TOKEN_TYPE_MAP.put("PTR", TokenType.PtrWord);
        NAME_TO_TOKEN_TYPE_MAP.put("BYTE", TokenType.ByteWord);
        NAME_TO_TOKEN_TYPE_MAP.put("EAX", TokenType.Reg32);
        NAME_TO_TOKEN_TYPE_MAP.put("EBX", TokenType.Reg32);
        NAME_TO_TOKEN_TYPE_MAP.put("DX", TokenType.Reg16);
        NAME_TO_TOKEN_TYPE_MAP.put("SI", TokenType.Reg16);
        NAME_TO_TOKEN_TYPE_MAP.put("ECX", TokenType.Reg32);
        NAME_TO_TOKEN_TYPE_MAP.put("BX", TokenType.Reg16);
        NAME_TO_TOKEN_TYPE_MAP.put("DI", TokenType.Reg16);
        NAME_TO_TOKEN_TYPE_MAP.put("BP", TokenType.Reg16);
        NAME_TO_TOKEN_TYPE_MAP.put("ES", TokenType.SegmentRegister);
        NAME_TO_TOKEN_TYPE_MAP.put("CS", TokenType.SegmentRegister);
        NAME_TO_TOKEN_TYPE_MAP.put("DS", TokenType.SegmentRegister);
        NAME_TO_TOKEN_TYPE_MAP.put("FS", TokenType.SegmentRegister);
    }

    private static Pattern numberHexRegex = Pattern.compile("^[0-9A-F]+H$");
    private static Pattern numberDecRegex = Pattern.compile("^[0-9]+$");
    private static Pattern numberBinRegex = Pattern.compile("^[01]+B$");
    private static Pattern identifierRegex = Pattern.compile("^[A-Z]\\w*$");

    public TokenType type;
    public String stringToken;

    public Token(String string) {
        stringToken = string;
        if (NAME_TO_TOKEN_TYPE_MAP.containsKey(string)) {
            type = NAME_TO_TOKEN_TYPE_MAP.get(string);
        } else if (numberHexRegex.matcher(string).matches()) {
            type = TokenType.HexNumber;
        } else if (numberDecRegex.matcher(string).matches()) {
            type = TokenType.DecNumber;
        } else if (numberBinRegex.matcher(string).matches()) {
            type = TokenType.BinNumber;
        } else if (identifierRegex.matcher(string).matches()) {
            type = TokenType.Identifier;
        } else if (string.trim().startsWith("\"") && string.trim().endsWith("\"")) {
            type = TokenType.Text;
        }

        if (type == TokenType.Text)
            stringToken = string.substring(1, string.length() - 1);
    }


    @Override
    public String toString() {
        return "Token{" +
                "Type=" + type +
                ", stringToken='" + stringToken + '\'' +
                '}';
    }
}
